src/bit-systems/link-hover-menu.ts

```diff
@@ -1,47 +1,150 @@
+const menuQuery = defineQuery([LinkHoverMenu]);
+const hoveredQuery = defineQuery([HoveredRemoteRight]);
 const clickedMenuItemQuery = defineQuery([Interacted, LinkHoverMenuItem]);
 
-export function linkHoverMenuSystem(world: HubsWorld) {
-  // Assumes always only single LinkHoverMenu entity exists for now.
-  // TODO: Take into account for more than one for VR
-  const menuEid = anyEntityWith(world, LinkHoverMenu)!;
-  const menuObject = world.eid2obj.get(menuEid)!;
-
-  // Save Link object eid in LinkHoverMenu when hovered.
-  hoveredEnterQuery(world).forEach((eid: number) => {
-    LinkHoverMenu.targetObjectRef[menuEid] = eid;
-    const targetObject = world.eid2obj.get(eid)!;
-    targetObject.add(menuObject);
-  });
+function updateLinkMenuTarget(world: HubsWorld, menu: EntityID, sceneIsFrozen: boolean) {
+  if (LinkHoverMenu.targetObjectRef[menu] && !entityExists(world, LinkHoverMenu.targetObjectRef[menu])) {
+    // Clear the invalid entity reference. (The link entity was removed).
+    LinkHoverMenu.targetObjectRef[menu] = 0;
+    return;
+  }
+
+  if (sceneIsFrozen) {
+    LinkHoverMenu.targetObjectRef[menu] = 0;
+    return;
+  }
+
+  const hovered = hoveredQuery(world);
+  const target = hovered.map(eid => findAncestorWithComponent(world, Link, eid))[0] || 0;
+  if (target) {
+    LinkHoverMenu.targetObjectRef[menu] = target;
+    LinkHoverMenu.clearTargetTimer[menu] = world.time.elapsed + 1000;
+    return;
+  }
+
+  if (hovered.some(eid => findAncestorWithComponent(world, LinkHoverMenu, eid))) {
+    LinkHoverMenu.clearTargetTimer[menu] = world.time.elapsed + 1000;
+    return;
+  }
+
+  if (world.time.elapsed > LinkHoverMenu.clearTargetTimer[menu]) {
+    LinkHoverMenu.targetObjectRef[menu] = 0;
+    return;
+  }
+}
+
+async function handleLinkClick(world: HubsWorld, button: EntityID) {
+  const exitImmersive = async () => await handleExitTo2DInterstitial(false, () => {}, true);
+
+  const menu = findAncestorWithComponent(world, LinkHoverMenu, button)!;
+  const linkEid = LinkHoverMenu.targetObjectRef[menu];
+  const src = APP.getString(Link.url[linkEid])!;
+  const url = new URL(src);
+  const linkType = Link.type[linkEid];
+  switch (linkType) {
+    case LinkType.LINK:
+      await exitImmersive();
+      window.open(src);
+      break;
+    case LinkType.AVATAR:
+      const avatarId = new URL(src).pathname.split("/").pop();
+      window.APP.store.update({ profile: { avatarId } });
+      APP.scene!.emit("avatar_updated");
+      break;
+    case LinkType.SCENE:
+      APP.scene!.emit("scene_media_selected", src);
+      break;
+    case LinkType.WAYPOINT:
+      // move to waypoint w/o writing to history
+      window.history.replaceState(null, "", window.location.href.split("#")[0] + url.hash);
+      break;
+    case LinkType.ROOM:
+      const waypoint = url.hash && url.hash.substring(1);
+      // move to new room without page load or entry flow
+      const hubId = hubIdFromUrl(url);
```

Comment:
Off topic from this PR: Question, do we ensure that url is valid somewhere? If not, we may add validations in inflators or somewhere else (as we discussed validations before).

Validations may be added in another PR.